### PR TITLE
move load/save file to a controller

### PR DESCRIPTION
**Pull request by schiesbn**
this fix the error response on failed load operation, introduced by this changes: https://github.com/owncloud/files_texteditor/pull/77/files

fix https://github.com/owncloud/core/issues/16264

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
@schiesbn travis is failing because it tries to build on php 5.3 - because I doubt that this app's master branch will work on less then 8.1 we can ignore this - but we should set the requirements properly in info.xml

File: appinfo/routes.php [hunk 1]
```diff
@@ -1,14 +1,39 @@
...
+namespace OCA\Files_Texteditor\AppInfo;
+
+$app = new Application();
+
+$app->registerRoutes($this, array('routes' => array(
+
+	[
+		'name' => 'FileHandling#load',
+		'url' => '/ajax/loadfile',
+		'verb' => 'POST'
```
Comment (DeepDiver1975) [inline appinfo/routes.php]: Load should be a GET command

File: appinfo/routes.php [hunk 2]
```diff
@@ -1,14 +1,39 @@
...
+	],
+	[
+		'name' => 'FileHandling#save',
+		'url' => '/ajax/savefile',
+		'verb' => 'POST'
```
Comment (DeepDiver1975) [inline appinfo/routes.php]: generally speaking PUT should fit better - but might require to much changes

**Comment by DeepDiver1975** [comment]
@schiesbn you see me honestly impressed by your changes here whit moving the test edtior to app framework - nice job! Congrats!

File: controller/filehandlingcontroller.php [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+					return new DataResponse(['mtime' => $newmtime, 'size' => $newsize], Http::STATUS_OK);
+				} else {
+					// Not writeable!
+					return new DataResponse([ 'message' => $this->l->t('Insufficient permissions')],
+						Http::STATUS_BAD_REQUEST);
+					$this->logger->error('User does not have permission to write to file: ' . $path,
+						['app' => 'files_texteditor']);
+				}
+			}
+		} else if($path == '') {
```
Comment (LukasReschke) [inline controller/filehandlingcontroller.php]: `===`

File: controller/filehandlingcontroller.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			return new DataResponse(['message' => $this->l->t('File path not supplied')], Http::STATUS_BAD_REQUEST);
+			$this->logger->error('No file path supplied');
+		} else if($mtime == '') {
```
Comment (LukasReschke) [inline controller/filehandlingcontroller.php]: `===`

File: controller/filehandlingcontroller.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	/**
+	 * load text file
+	 *
+	 * @NoAdminRequired
+	 *
+	 * @param string $dir
+	 * @param string $filename
+	 * @return JSONResponse
```
Comment (LukasReschke) [inline controller/filehandlingcontroller.php]: Sure?

File: controller/filehandlingcontroller.php [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	 * @NoAdminRequired
+	 *
+	 * @param $path
+	 * @param $filecontents
+	 * @param $mtime
+	 * @return DataResponse|BadRequest
+	 */
+	public function save($path, $filecontents, $mtime) {
+
+		if($path != '' && $mtime != '') {
```
Comment (LukasReschke) [inline controller/filehandlingcontroller.php]: `!==`

File: controller/filehandlingcontroller.php [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	/**
+	 * save text file
+	 *
+	 * @NoAdminRequired
+	 *
+	 * @param $path
+	 * @param $filecontents
+	 * @param $mtime
```
Comment (LukasReschke) [inline controller/filehandlingcontroller.php]: Type hints :)

File: controller/filehandlingcontroller.php [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	 * @return DataResponse|BadRequest
+	 */
+	public function save($path, $filecontents, $mtime) {
+
+		if($path != '' && $mtime != '') {
+			// Get file mtime
+			$filemtime = $this->view->filemtime($path);
+			if($mtime != $filemtime) {
```
Comment (LukasReschke) [inline controller/filehandlingcontroller.php]: !==

File: controller/filehandlingcontroller.php [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	private $logger;
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * @param string $AppName
+	 * @param IRequest $request
+	 * @param IL10N $l10n
+	 * @param View $view
+	 * @paran ILogger $logger
```
Comment (LukasReschke) [inline controller/filehandlingcontroller.php]: Typo

**Comment by LukasReschke** [comment]
:rocket: :beer: :dancer:

**Comment by DeepDiver1975** [comment]
@schiesbn let me know as soon as unit tests have been added - we can then adopt travis.yml to get them executed

**Comment by schiesbn** [comment]
I tried to keep the changes as small as possible. That's why I copy&pasted as much as possible. But you are both right about your comments, I will integrate them tomorrow together with the unit tests

**Comment by schiesbn** [comment]
All comments are addressed and unit tests are added... Ready for a final review.

**Comment by DeepDiver1975** [comment]
> unit tests are added

Okay - let's add travis unit test execution ....

**Comment by DeepDiver1975** [comment]
travis is integrated

**Comment by schiesbn** [comment]
@DeepDiver1975 Thanks! And all tests already passed :smile:

**Comment by nickvergessen** [comment]
Works nice so far.
the only thing I think needs some more work is:

1. Create a file.txt
2. Open 2nd tab and delete the file there
3. In tab1 click on the file to edit it

### Expected

A note that the file was deleted

### Actual

A read only empty editor is opened

----

Is this related to your work @schiesbn ?

**Comment by schiesbn** [comment]
@nickvergessen I think also the old code didn't catch this case. But I added another check so you should get now a proper error message.

File: controller/filehandlingcontroller.php [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	 */
+	public function load($dir, $filename) {
+		try {
+			if (!empty($filename)) {
+				$path = $dir . '/' . $filename;
+				$writable = $this->view->isUpdatable($path);
+				$mime = $this->view->getMimeType($path);
+				$mtime = $this->view->filemtime($path);
+				$filecontents = $this->view->file_get_contents($path);
+				if ($filecontents) {
```
Comment (nickvergessen) [inline controller/filehandlingcontroller.php]: Now you can not edit files anymore that are empty, or `0`

Suggestion, replace the code until here with:
```php
		try {
			if (!empty($filename)) {
				$path = $dir . '/' . $filename;
				if ($this->view->file_exists($path)) {
					$writable = $this->view->isUpdatable($path);
					$mime = $this->view->getMimeType($path);
					$mtime = $this->view->filemtime($path);
					$filecontents = $this->view->file_get_contents($path);
```

Comment (nickvergessen) [inline controller/filehandlingcontroller.php]: will commit it with another problem I just found:
The page title gets screwed up in case of an error.

**Comment by nickvergessen** [comment]
@DeepDiver1975 @schiesbn I fixed the whitelist paths, added js/vendor to scrutinizers ignore and fixed the test so it gets actually executed.

:+1: from me

**Comment by nickvergessen** [comment]
Added a new test and fixed it to also check the result of `file_get_content` against `=== false`

**Comment by davitol** [comment]
:+1:  An error "Could not obtain lock type 1 on "/opt/owncloud/data/admin/files/asdasda.txt"." is shown now when trying to edit a deleted file.
owncloud/core#16264

**Comment by schiesbn** [comment]
Thanks for the improvements @nickvergessen... Looks good! :+1: for your changes.

**Pull request merged**